lib/MusicBrainz/LWP/TimelyUserAgent.pm

```diff
@@ -0,0 +1,110 @@
+package MusicBrainz::LWP::TimelyUserAgent;
+
+use Moose;
+
+use DBDefs;
+use HTTP::Status qw( HTTP_SERVICE_UNAVAILABLE );
+use MusicBrainz::Server::Exceptions qw( GenericTimeout );
+use POSIX qw( SIGALRM );
+use Try::Tiny;
+
+extends 'LWP::UserAgent';
+
+use MooseX::Types::Moose qw( Int );
+
+has 'sending_timeout' => (
+    is => 'rw',
+    isa => 'Int',
+);
+
+sub new {
+    my ($class, %args) = @_;
+
+    $args{agent} = DBDefs->LWP_USER_AGENT unless exists $args{agent};
+
+    my $sending_timeout =
+        exists $args{sending_timeout}
+            ? delete $args{sending_timeout}
+            : 1;
+
+    my $obj = $class->SUPER::new(%args);
+
+    return $class->meta->new_object(
+        __INSTANCE__ => $obj,
+        sending_timeout => $sending_timeout,
+    );
+}
+
+around send_request => sub {
+    my ($orig, $self, @args) = @_;
+
+    my $sending_timeout = $self->sending_timeout // 0;
+    my $request_timeout = $self->timeout // 0;
+
+    return $self->$orig(@args) if $sending_timeout <= 0 || $request_timeout <= 0;
+
+    my ($request, $arg, $size) = @args;
+
+    my $action_timeout = $sending_timeout + $request_timeout;
+
+    alarm($action_timeout);
+
+    my $action = POSIX::SigAction->new(sub {
+        MusicBrainz::Server::Exceptions::GenericTimeout->throw(
+            $request->method . ' ' . $request->uri .
+            " took more than $action_timeout seconds (sending included)"
+        );
+    });
+    $action->safe(1);
+    POSIX::sigaction(SIGALRM, $action);
+
+    my $response;
+    try {
+        $response = $self->$orig(@args);
+    } catch {
+        my $err = $_;
+        if (ref($err) eq 'MusicBrainz::Server::Exceptions::GenericTimeout') {
+            $response = $self->SUPER::_new_response(
+                $request,
+                HTTP_SERVICE_UNAVAILABLE,
+                $err->message,
+                $err->as_string,
+            );
+        } else {
+            die $err;
+        }
+    };
+
+    alarm(0);
+
+    return $response;
+};
+
+__PACKAGE__->meta->make_immutable( inline_constructor => 0 );
+no Moose;
+1;
+
+=head1 NAME
+
+MusicBrainz::LWP::TimelyUserAgent - Web user agent class with sending timeout
+
+=head1 DESCRIPTION
+
+If both C<sending_timeout> and C<timeout> are positive integers,
+when sending a request and receiving a response takes more than
+C<sending_timeout> plus C<timeout> seconds, it forges a response
```

Comment:
The goal is to get around `LWP::UserAgent`’s limitation which may hang indefinitely before getting connected, that is exactly what happened when we experienced networking issues.

The `sending_timeout` just gives an extra second so as to allow to distinguish between timeout correctly handled by `LWP::UserAgent` and timeout saved by `TimelyUserAgent`. It doesn’t attempt to take processing time, redirect time, and so on, into account.

Turning it into a really global timeout (on returning the final response) would require:
* either to wrap `request` subroutine which is recursive (not alarm friendly),
* or to wrap `get`, `post`, and `mirror` subs and replace `request` calls with `post` calls in `::Controller::Discourse` and `::Test::HTML5`.

I’d still prefer the current solution, maybe just explain it better in POD.

Parameters specification will follow, once we agree on which issue we try to address.

P.S. Isn’t using `alarm` here going to interfere with `alarm` call in `MusicBrainz::Server`?

Reply:
> P.S. Isn’t using alarm here going to interfere with alarm call in MusicBrainz::Server?

Yeah...good point. We should probably be using `AnyEvent` for this, though we'll have to pull it in as a dependency. I played around with it and it looks like there are workable solutions that way:

```Perl
# cpanm AnyEvent AnyEvent::HTTP::LWP::UserAgent EV

BEGIN { $ENV{PERL_ANYEVENT_MODEL} = 'EV'; }

use AnyEvent;
use AnyEvent::HTTP::LWP::UserAgent;
use Data::Dumper;

my $ua = AnyEvent::HTTP::LWP::UserAgent->new;
my $cv = AE::cv;
$cv->begin;

my $response;
my $timeout = 0.5; # seconds
my $got_timeout = 0;

my $req_cv = $ua->get_async('https://musicbrainz.org/');

$req_cv->cb(sub {
    $response = shift->recv;
    $cv->end;
});

my $timer = AnyEvent->timer(after => $timeout, cb => sub {
    $req_cv->croak;
    $got_timeout = 1;
    $cv->croak;
});

$cv->recv;

undef $timer; # cancels the timer

if ($got_timeout) {
    print 'Got timeout';
} else {
    print Dumper($response);
}
```

This would probably have to be implemented as a utility function rather than an LWP subclass.